File: leetcode-go/Q001-Q050/Q039-Combination-Sum/Q039-Combination-Sum.go
package Q039_Combination_Sum

import "sort"

func combinationSum(candidates []int, target int) [][]int {
	ans := [][]int{}
	sort.Ints(candidates)
	dfs(&ans, candidates, []int{}, target, 0)
	return ans
}

func dfs(ans *[][]int, candidates, combs []int, target int, begin int) {
	if target == 0 {
		*ans = append(*ans, combs)
		return
	}
	for i := begin; i < len(candidates) && target >= candidates[i]; i++ {
		ncombs := make([]int, len(combs))
		copy(ncombs, combs)

		dfs(ans, candidates, append(ncombs, candidates[i]), target-candidates[i], i)
	}
}

File: leetcode-java/src/archive/backtracking/Q039_Combination_Sum.java
package archive.backtracking;

import java.util.*;

public class Q039_Combination_Sum {

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> result = new LinkedList<>();
        dfs(result, new LinkedList<>(), candidates, target, 0);
        return result;
    }

    private void dfs(List<List<Integer>> result, List<Integer> list, int[] nums, int target, int pos) {
        if (target == 0) {
            result.add(new LinkedList<>(list));
            return;
        }
        for (int i = pos; i < nums.length; i++) {
            if (target-nums[i] < 0) {
                continue;
            }
            list.add(nums[i]);
            dfs(result, list, nums, target-nums[i], i);
            list.remove(list.size()-1);
        }
    }

}
